README.md
# fauxmo-plugins

master: [![master branch build status](https://travis-ci.org/n8henrie/fauxmo-plugins.svg?branch=master)](https://travis-ci.org/n8henrie/fauxmo-plugins/branches)
dev: [![dev branch build status](https://travis-ci.org/n8henrie/fauxmo-plugins.svg?branch=dev)](https://travis-ci.org/n8henrie/fauxmo-plugins/branches)

Plugins for Fauxmo (emulated Wemo devices for the Amazon Echo)

## Introduction and Rationale

As of Fauxmo 0.4.0, I am splitting the plugins out of the main
[Fauxmo](https://github.com/n8henrie/fauxmo) code. Plugins were previously
referred to as `handlers`.

- I've implemented a crude plugin import system that allows users to create
  and user their own plugins by inheriting from `FauxmoPlugin` and including
  the path to the file in their config.
- This means I can remove dependencies from Fauxmo that some users may not use
  (e.g. my `homeassistant.remote` stuff), and hopefully users unfamiliar with
  virtualenvs will quit breaking their installations due to my pinned versions.
- I'll be able to update plugins without having to release a new version of
  Fauxmo core, keeping them in their own separate VCS silos.
- I can implement probably unsafe code like the new `CommandLinePlugin` as a
  plugin, and users wanting such functionality can easily install at their own
  risk without making *all* Fauxmo users do so.

## Using plugins

### Personal user plugins

The biggest motivation for the changes to Fauxmo 0.4.0 was to allow users to
create Fauxmo plugins to scratch their own itch. There are only a few
requirements to get started:

- Your plugin will be class that inherits from the
  `fauxmo.plugins.FauxmoPlugin` ABC.
- Your plugin will override the abstract methods `on`, `off`, and `get_state`,
  which will unsurprisingly be called when you tell Alexa to turn a Fauxmo
  device on or off, and when Alexa [queries the device
  state](https://github.com/n8henrie/fauxmo-plugins/issues/3). The `on` and
  `off` methods should return a `bool` that suggests whether they succeeded,
  and `get_state` should return `"on"`, `"off"`, or `"unknown"`. If you have no
  way to query state, consider using a simple `return "unknown"` as your
  `get_state` method.
- Your plugin will be initialized if the exact (case sensitive) name is listed
  as a key under the `PLUGINS` section in your Fauxmo configuration (please see
  the Fauxmo docs for details), and if you include the path to the file
  that provides your plugin class as the `path` subkey.
- Each device you plan to use with a plugin will be listed under the `DEVICES`
  key of your plugin.
- Your plugin class will receive several `kwargs` on initialization, including
  the `name` and `port` that the Fauxmo device will use, as defined in your
  Fauxmo config. Hopefully this will allow users to implement some custom
  debugging / logging features.
- If you want your plugin class to determine its own port in code or if you
  decide to override the `__init__` method, you may need to do some combination
  of the following:
  - `super().__init__(name=name, port=port)` in your custom `__init__`.
  - Define the "private" `self._port` attribute.
  - Override the `FauxmoPlugin.port()` property.
- If your plugin has external dependencies, I highly recommend that you include
  the version of the dependency in your module-level docstring, especially if
  you're going to publish your plugin as a Gist.

I will not be providing much support to users needing help with plugins. I'm an
intermediate Pythonista at best, so if you are just learning about classes,
inheritance, and abstract methods, please feel free to make a new issue for
problems you encounter, but I might not be much help. On the other hand, if
you're a more advanced Pythonista, feel free to make suggestions.

### Plugins provided in the `fauxmo-plugins` repo

I'm going to include a few handy plugins here for reference. To use them in
your Fauxmo installation, all you need to do is get a local copy of the file
and include the `path` in your Fauxmo config as described above. You could do
this a few ways:

1. Clone the repo `git clone https://github.com/n8henrie/fauxmo-plugins.git`
1. Use `"path": "~/path/to/fauxmo-plugins/exampleplugin.py"` in your config

Alternatively,

1. Download the specific plugin you're interested in by clicking on the file in
   your web browser, clicking the `Raw` button, and using `wget` or `curl` to
   download the resulting file.
1. Include the path to that file as the `path` for that plugin.

If you think your plugin would be good to include in the `fauxmo-plugins`
repo, feel free to send me a pull request. To be merged:

- Must by python 3.6+ compatible
- Include a reasonable docstring that:
  - Explains the intended purpose, usage, and any required config variables.
  - Includes as pinned version numbers for any dependencies at the end.
- The file should:
  - Include type annotations.
  - Pass `mypy --ignore-missing-imports`.
  - Pass `flake8`.

Additionally, I'd like to include a list of interesting plugins here in the
README, even if the owners don't want to be included in this repo.

### Pre-installed plugins

Fauxmo comes with few plugin classes already available in the `fauxmo.plugins`
package. If you think your plugin would be good to include as one of these,
send a PR to that repo. Please note that I would strongly prefer not to have
any 3rd party / non-stdlib dependencies for the core Fauxmo package, in
addition to the requirements for the `fauxmo-plugin` repo above.

## Tests

Tests are *highly* recommended.

The modules are not installable, so you'll need to be able to import your class
directly from the module file. This can be done by monkeypatching `sys.path`,
but I prefer if you just use `python3 -m pytest tests/` from the root
directory, which will prompt Python to add it to `PYTHONPATH`, and should allow
you to import your class: `from myplugin import MyPlugin`.

You may find the `fauxmo_server` pytest fixture helpful -- given the path to a
config file as its only argument, it returns a context manager that is a Fauxmo
instance using that config. Using this context manager and a small sample
config, you can simulate receiving a "turn on" command from the Echo by posting
`'<BinaryState>1</BinaryState>'` to the
`http://localhost:12345/upnp/control/basicevent1` endpoint, where `localhost`
is your config's `["FAUXMO"]["ip_address"]` and `12345` is the `port` for one
of your plugin's Fauxmo device instances. See `tests/test_restapiplugin.py` and
`tests/test_restapiplugin_config.json` for an example.

Alternatively, you can test your plugin directly without using the rest of the
Fauxmo machinery by reading in your config and ensuring that e.g.
`YourPlugin(**device).on()` works as intended (again, see
`tests/test_restapiplugin.py` as an example).

NB: To facilitate simultaneous development between n8henrie/fauxmo and
n8henrie/fauxmo-plugins, `requirements-test.txt` pulls the latest version of
the dev branch of Fauxmo from GitHub; this is to help ensure that the automated
CI tests can run if I am pushing out changes to both repos that are not yet
ready to be merged into master or get pushed to PyPI. I suggest that if you're
developing locally and need to make simultaneous PRs against Fauxmo and
Fauxmo-plugins that you check out a separate branch in each and temporarily
change the `requirements-test.txt` file to something like `-e
git+file:///abs/path/to/local/fauxmo/repo#egg=fauxmo`.

## Interesting Plugins (not included in this repo)

- None yet. Why not be the first?

## Troubleshooting / FAQ

See also: [Fauxmo FAQ](https://github.com/n8henrie/fauxmo#troubleshooting--faq)


CHANGELOG.md
# [Changelog](https://keepachangelog.com)

## 20190702

- Major changes to HomeAssistantPlugin thanks to
  [jld3103](https://github.com/jld3103):
  - Renamed plugin file
  - Now uses `Authorization: Bearer TOKEN` for authentication
  - No longer has any non-stdlib dependencies, and as such, I will likely move
    this into the main Fauxmo repo and have it installed automatically
    alongside the other default plugins.
- Format with [black](https://github.com/python/black)

## 20180508

- Decided to keep a changelog
- Added MQTT Plugin thanks to [Perforex](https://github.com/Perforex)


tests/conftest.py
"""conftest.py :: Setup fixtures for pytest."""

import time
from multiprocessing import Process
from types import TracebackType
from typing import Callable, Optional, Type

import pytest
from fauxmo import fauxmo
from fauxmo.utils import get_local_ip


class TestFauxmoServer:
    """Runs Fauxmo in a separate thread."""

    def __init__(self, config_path_str: str) -> None:
        """Initialize test Fauxmo server with path to config."""
        self.config_path_str = config_path_str

    def __enter__(self) -> str:
        """Start a TextFauxmoServer, returns the ip address it's running on."""
        self.server = Process(
            target=fauxmo.main,
            kwargs={"config_path_str": self.config_path_str},
            daemon=True,
        )
        self.server.start()
        time.sleep(1)
        return get_local_ip()

    def __exit__(
        self,
        exc_type: Optional[Type[BaseException]],
        exc_value: Optional[Exception],
        traceback: Optional[TracebackType],
    ) -> None:
        """Terminate the server and join the thread on exit."""
        self.server.terminate()
        self.server.join()


@pytest.fixture(scope="module")
def fauxmo_server() -> Callable[[str], TestFauxmoServer]:
    """Use a pytest fixture to provide the TestFauxmoServer context manager."""
    return TestFauxmoServer


tests/test_mqttplugin.py
"""test_restapiplugin.py :: Tests for Fauxmo's `RESTAPIPlugin`."""

import json
import time
from unittest.mock import MagicMock, patch

from mqttplugin import MQTTPlugin

config_path_str = "tests/test_mqttplugin_config.json"


def test_mqttplugin_mosquitto_dot_org() -> None:
    """Test MQTTPlugin against test.mosquitto.org."""
    with open(config_path_str) as f:
        config: dict = json.load(f)

    mosquitto_devices = (
        device
        for device in config["PLUGINS"]["MQTTPlugin"]["DEVICES"]
        if device["mqtt_server"] == "test.mosquitto.org"
    )
    for device_conf in mosquitto_devices:
        device = MQTTPlugin(**device_conf)

        for _ in range(100):
            if device.subscribed is True:
                break
            time.sleep(0.1)
        else:
            assert False, "Time out waiting for subscribe."

        assert device.on() is True
        for _ in range(20):
            state = device.get_state()
            if state != "unknown":
                break
            time.sleep(0.1)
        assert state == "on"

        assert device.off() is True
        for _ in range(20):
            state = device.get_state()
            if state != "on":
                break
            time.sleep(0.1)
        assert state == "off"


@patch("mqttplugin.Client", autospec=True)
def test_mqtt_auth(mock_client: MagicMock) -> None:
    """Ensure auth is being used if available."""
    mock_instance = mock_client.return_value
    with open(config_path_str) as f:
        config: dict = json.load(f)

    device_conf = next(
        device
        for device in config["PLUGINS"]["MQTTPlugin"]["DEVICES"]
        if device["mqtt_server"] == "mqtt.yes_auth.no_state"
    )
    MQTTPlugin(**device_conf)
    mock_instance.username_pw_set.assert_called_once_with(
        "MyUser", "MyPassword"
    )


@patch("mqttplugin.Client", autospec=True)
def test_mqtt_nostate(mock_client: MagicMock) -> None:
    """If state_cmd is not specified, loop_start is not called."""
    mock_instance = mock_client.return_value
    with open(config_path_str) as f:
        config: dict = json.load(f)

    device_conf = next(
        device
        for device in config["PLUGINS"]["MQTTPlugin"]["DEVICES"]
        if device["mqtt_server"] == "mqtt.yes_auth.no_state"
    )
    device = MQTTPlugin(**device_conf)
    mock_instance.loop_start.assert_not_called()
    mock_instance.subscribe.assert_not_called()
    assert device.subscribed is False


@patch("mqttplugin.Client", autospec=True)
def test_mqtt_noauth(mock_client: MagicMock) -> None:
    """Ensure auth is not being used if not configured."""
    mock_instance = mock_client.return_value
    with open(config_path_str) as f:
        config: dict = json.load(f)

    device_conf = next(
        device
        for device in config["PLUGINS"]["MQTTPlugin"]["DEVICES"]
        if device["mqtt_server"] == "mqtt.no_auth.yes_state"
    )
    MQTTPlugin(**device_conf)
    mock_instance.username_pw_set.assert_not_called()


tests/test_restapiplugin.py
"""test_restapiplugin.py :: Tests for Fauxmo's `RESTAPIPlugin`."""

import json
import time
from multiprocessing import Process
from typing import Generator

import httpbin
import pytest
import requests
from restapiplugin import RESTAPIPlugin


config_path_str = "tests/test_restapiplugin_config.json"


@pytest.fixture(scope="function")
def restapiplugin_target() -> Generator:
    """Simulate the endpoints triggered by RESTAPIPlugin."""
    fauxmo_device = Process(
        target=httpbin.core.app.run,
        kwargs={"host": "127.0.0.1", "port": 8000},
        daemon=True,
    )

    fauxmo_device.start()
    time.sleep(1)

    yield

    fauxmo_device.terminate()
    fauxmo_device.join()


def test_restapiplugin_integration(
    fauxmo_server: pytest.fixture, restapiplugin_target: pytest.fixture
) -> None:
    """Test "on" and "off" actions for RESTAPIPlugin.

    This test uses requests to `post` a value to a Fauxmo device that
    simulates the way the Echo interacts with the Fauxmo server when it gets a
    request to turn something `on` or `off`.

    It requires both the Fauxmo server be running as well as a "target" HTTP
    server for Fauxmo to interact with.

    requests.post -> Fauxmo device running at `port` -> target url (httpbin
    in this case, from the device's `on_cmd`)
    """
    command_format = (
        "SOAPACTION: " '"urn:Belkin:service:basicevent:1#{}BinaryState"'.format
    )
    data_template = "<BinaryState>{}</BinaryState>".format

    data_get_state = command_format("Get")
    data_on = command_format("Set") + data_template(1)
    data_off = command_format("Set") + data_template(0)

    with fauxmo_server(config_path_str) as fauxmo_ip:
        base_url = f"http://{fauxmo_ip}:12345/upnp/control/basicevent1"
        resp_on = requests.post(base_url, data=data_on.encode())
        resp_off = requests.post(base_url, data=data_off.encode())
        resp_state = requests.post(base_url, data=data_get_state.encode())

    assert resp_on.status_code == 200
    assert resp_off.status_code == 200
    assert resp_state.status_code == 200


def test_restapiplugin_unit(restapiplugin_target: pytest.fixture) -> None:
    """Test simple unit tests on just the device without the integration."""
    with open(config_path_str) as f:
        config: dict = json.load(f)

    for device_conf in config["PLUGINS"]["RESTAPIPlugin"]["DEVICES"]:
        device = RESTAPIPlugin(**device_conf)
        assert device.on() is True
        assert device.off() is True

        state = device.get_state()
        if device.state_cmd is not None:
            assert state == "on"
        else:
            assert state == "unknown"


mqttplugin.py
"""Fauxmo plugin for simple MQTT requests.

The on and off methods publish a value to the given MQTT queue. The get_status
method subscribes to an MQTT queue to asynchronously receive the status
published from the device. If the device doesn’t publish a status via MQTT then
omit the state_cmd and the plugin will return a status of “unknown”. The status
received from the device is passed back unchanged to fauxmo which is expecting
“on”, “off” or “unknown”.

Behaviour of the plugin is not entirely predictable if mqttserver and mqttport
are omitted, please make sure to set in your config.

Example config:

```
{
    "FAUXMO": {
        "ip_address": "auto"

    },
    "PLUGINS": {
        "MQTTPlugin": {
            "path": "/path/to/mqttplugin.py",
            "DEVICES": [
                {
                    "port": 12349,
                    "on_cmd": [ "Home/Light/Study01", "1" ],
                    "off_cmd": [ "Home/Light/Study01", "0" ],
                    "state_cmd": "Home/Light/Study01",
                    "name":"MQTT Study Light 1",
                    "mqtt_server": "test.mosquitto.org",
                    "mqtt_port": 1883
                },
                {
                    "port": 12350,
                    "on_cmd": [ "Home/Light/Study02", "1" ],
                    "off_cmd": [ "Home/Light/Study02", "0" ],
                    "state_cmd": "Home/Light/Study02",
                    "name":"MQTT Study Light 2",
                    "mqtt_server": "test.mosquitto.org",
                    "mqtt_port": 1883
                },
                {
                    "port": 12351,
                    "on_cmd": [
                        "home-assistant/devices/cmnd/sonoff1/POWER",
                        "ON"
                    ],
                    "off_cmd": [
                        "home-assistant/devices/cmnd/sonoff1/POWER",
                        "OFF"
                    ],
                    "state_cmd": "home-assistant/devices/stat/sonoff1/POWER",
                    "name":"Hass MQTT device",
                    "mqtt_server": "192.168.1.108",
                    "mqtt_port": 1883,
                    "mqtt_user": "MyHassUser",
                    "mqtt_pw": "MySecretP@ssword"
                }
            ]
        }
    }
}
```

Dependencies:
    paho-mqtt==1.3.1
"""

from typing import List, Sequence

from fauxmo.plugins import FauxmoPlugin
from paho.mqtt.client import Client, MQTTMessage


class MQTTPlugin(FauxmoPlugin):
    """Fauxmo plugin to interact with an MQTT server by way of paho."""

    def __init__(
        self,
        *,
        name: str,
        port: int,
        off_cmd: Sequence[str],
        on_cmd: Sequence[str],
        mqtt_port: int = 1883,
        mqtt_pw: str = None,
        mqtt_user: str = None,
        mqtt_server: str = "127.0.0.1",
        state_cmd: str = None,
    ) -> None:
        """Initialize an MQTTPlugin instance.

        Kwargs:
            name: device name
            port: Port for Fauxmo to make this device avail to Echo

            mqttport: MQTT server port
            mqttserver: MQTT server address
            mqttuser: MQTT username
            mqttpw: MQTT password
            off_cmd: [ MQTT Queue, value to be publshed as str ] to turn off
            on_cmd: [ MQTT Queue, value to be publshed as str ] to turn on
            state_cmd: MQTT Queue to get state
        """
        self.on_cmd, self.on_value = on_cmd[0], on_cmd[1]
        self.off_cmd, self.off_value = off_cmd[0], off_cmd[1]
        self.state_cmd = state_cmd
        self.status = "unknown"
        self._subscribed = False

        self.client = Client()
        if mqtt_user or mqtt_pw:
            self.client.username_pw_set(mqtt_user, mqtt_pw)
        self.client.on_connect = self.on_connect
        self.client.on_subscribe = self.on_subscribe
        self.client.on_message = self.on_message

        self.client.connect(mqtt_server, mqtt_port, 60)

        super().__init__(name=name, port=port)

        # Looping thread only seems necessary for status updates
        if self.state_cmd is not None:
            self.client.loop_start()

    @property
    def subscribed(self) -> bool:
        """Property to return whether the subscription has completed."""
        return self._subscribed

    def on_subscribe(
        self, client: Client, userdata: str, mid: int, granted_qos: List[int]
    ) -> None:
        """Set attribute to show that initial subscription is complete."""
        self._subscribed = True

    def on_connect(
        self, client: Client, userdata: str, flags: dict, rc: int
    ) -> None:
        """Subscribe to state command on connect (or reconnect)."""
        if self.state_cmd is not None:
            self.client.subscribe(self.state_cmd)

    def on_message(
        self, client: Client, userdata: str, message: MQTTMessage
    ) -> None:
        """Process an incoming message."""
        status = message.payload.decode("utf-8")

        if status == self.off_value:
            self.status = "off"
        elif status == self.on_value:
            self.status = "on"

    def _publish(self, topic: str, value: str) -> bool:
        msg = self.client.publish(topic, value)
        try:
            msg.wait_for_publish()
        except ValueError:
            return False
        return True

    def on(self) -> bool:
        """Turn on MQTT device.

        Returns:
            True if device seems to have been turned on.

        """
        return self._publish(self.on_cmd, self.on_value)

    def off(self) -> bool:
        """Turn off MQTT device.

        Returns:
            True if device seems to have been turned off.

        """
        return self._publish(self.off_cmd, self.off_value)

    def get_state(self) -> str:
        """Return the self.status attribute.

        `self.status` is set asynchronously in on_message, so it may not
        immediately reflect state changed.

        Returns:
            State if known, else "unknown".

        """
        if self.state_cmd is None:
            return "unknown"

        return self.status


hassapiplugin.py
"""Fauxmo plugin to interact with HomeAssistant (hass) devices.

Install to Fauxmo by downloading or cloning and including in your Fauxmo
config. NB: For all Fauxmo plugins you must **manually** install any
required 3rd party modules (e.g. `homeassistant`) -- make sure you install
them where Fauxmo can find them (i.e. the same virtualenv).

If using the new hass authentication system instead of legacy api key, make
sure you have `api:` enabled in your hass config, and use a long-lived token
hass your `hass_password`.

Example config:

    def __init__(self, name: str, port: int, hass_host: str, hass_password:
                 str, entity_id: str, hass_port: int=8123) -> None:
```
{
    "FAUXMO": {
        "ip_address": "auto"
    },
    "PLUGINS": {
        "HassAPIPlugin": {
            "hass_host": "192.168.0.50",
            "hass_port": 8123,
            "hass_password": "abc123",
            "path": "/path/to/hassapiplugin.py",
            "DEVICES": [
                {
                    "name": "example hass device 1",
                    "port": 12345,
                    "entity_id": "switch.my_fake_switch"
                },
                {
                    "name": "example hass device 2",
                    "port": 12345,
                    "entity_id": "cover.my_fake_cover"
                }
            ]
        }
    }
}
```

Dependencies:
    homeassistant==0.57.3
"""

from collections import defaultdict
from typing import Dict

import homeassistant.remote
from fauxmo.plugins import FauxmoPlugin
from homeassistant.const import (  # noqa
        SERVICE_TURN_ON, SERVICE_TURN_OFF,
        SERVICE_OPEN_COVER, SERVICE_CLOSE_COVER,
        STATE_OFF, STATE_ON,
        STATE_CLOSED, STATE_OPEN,
        )


class HassAPIPlugin(FauxmoPlugin):
    """Fauxmo Plugin for HomeAssistant (hass) Python API.

    Allows users to specify Home Assistant services in their config.json and
    toggle these with the Echo. While this can be done with Home Assistant's
    REST API as well (example included), I find it easier to use the Python
    API.
    """

    service_map: Dict[str, Dict[str, str]] = defaultdict(dict)
    service_map.update({
            'cover': {
                'on': SERVICE_OPEN_COVER,
                'off': SERVICE_CLOSE_COVER,
                'on_state': STATE_OPEN,
                'off_state': STATE_CLOSED,
                },
            })

    def __init__(self, name: str, port: int, hass_host: str, entity_id: str,
                 hass_password: str=None, hass_port: int=8123) -> None:
        """Initialize a HassAPIPlugin instance.

        Args:
            hass_host: IP address of device running Home Assistant
            password: Home Assistant password
            entity_id: `entity_id` used by hass, one easy way to find is to
                       curl and grep the REST API, eg:
                          `curl http://IP/api/bootstrap | grep entity_id`
            hass_port: Port running hass on the host computer (default 8123)
        """
        self.hass_host = hass_host
        self.hass_password = hass_password
        self.entity_id = entity_id
        self.hass_port = hass_port

        self.domain = self.entity_id.split(".")[0]
        if self.domain == 'group':
            self.domain = 'homeassistant'
        self.api = homeassistant.remote.API(host=self.hass_host,
                                            api_password=self.hass_password,
                                            port=self.hass_port)

        super().__init__(name=name, port=port)

    def send(self, signal: str) -> bool:
        """Send a signal to the hass `call_service` function, returns True.

        The hass Python API doesn't appear to return anything with this
        function, but will raise an exception if things didn't seem to work, so
        I have it set to just return True, hoping for an exception if there was
        a problem.

        Args:
            signal (const): signal imported from homeassistant.const. I have
            imported SERVICE_TURN_ON and SERVICE_TURN_OFF, make sure you import
            any others that you need.
        """
        homeassistant.remote.call_service(self.api, self.domain, signal,
                                          {'entity_id': self.entity_id})
        return True

    def on(self) -> bool:
        """Turn the hass device on.

        Returns:
            Whether the device seems to have been turned on.

        """
        on_cmd = (
                HassAPIPlugin
                .service_map[self.domain.lower()]
                .get('on', SERVICE_TURN_ON)
                 )
        return self.send(on_cmd)

    def off(self) -> bool:
        """Turn the hass device off.

        Returns:
            Whether the device seems to have been turned off.

        """
        off_cmd = (
                HassAPIPlugin
                .service_map[self.domain.lower()]
                .get('off', SERVICE_TURN_OFF)
                 )
        return self.send(off_cmd)

    def get_state(self) -> str:
        """Query the state of the hass device."""
        smap = HassAPIPlugin.service_map[self.domain.lower()]
        response = homeassistant.remote.get_state(self.api, self.entity_id)

        # If no custom `on_state` key is set default to `STATE_ON`
        if response.state == smap.get('on_state', STATE_ON):
            return 'on'
        elif response.state == smap.get('off_state', STATE_OFF):
            return 'off'
        return 'unknown'


restapiplugin.py
r"""Fauxmo plugin that provides access to services exposing a REST API.

Uses `requests` for a simpler api.

Example config:

```
{
  "FAUXMO": {
    "ip_address": "auto"
  },
  "PLUGINS": {
    "RESTAPIPlugin": {
      "path": "/path/to/restapiplugin.py",
      "DEVICES": [
        {
          "port": 12340,
          "on_cmd": "http://192.168.1.3/myserver/switches/1/on",
          "off_cmd": "http://192.168.1.3/myserver/switches/1/off",
          "method": "GET",
          "name": "fake switch one",
          "state_cmd": "http://192.168.1.3/myserver/switches/1/state",
          "state_method": "GET",
          "state_response_on": "<p>Your switch is on."</p>"
          "state_response_off": "<p>Your switch is off.</p>"
        },
        {
          "port": 12341,
          "on_cmd": "http://localhost:54321/devices/garage%20light",
          "off_cmd": "http://localhost:54321/devices/garage%20light",
          "on_data": {"isOn": 1},
          "off_data": {"isOn": 0},
          "user": "this",
          "password": "that",
          "method": "PUT",
          "name": "fake Indigo switch"
        },
        {
          "name": "fake HomeAssistant switch via REST API",
          "port": 12342,
          "on_cmd": "http://192.168.0.5:8123/api/services/switch/turn_on",
          "off_cmd": "http://192.168.0.5:8123/api/services/switch/turn_off",
          "method": "POST",
          "headers": {
            "Authorization": "Bearer TOKENGOESHERE",
            "Content-Type": "application/json"
          },
          "on_json": {
            "entity_id": "switch.fake_hass_switch"
          },
          "off_json": {
            "entity_id": "switch.fake_hass_switch"
          },
          "state_method": "GET",
          "state_cmd":
            "http://192.168.0.5:8123/api/states/switch.fake_hass_switch",
          "state_response_on": "\"state\": \"on\"",
          "state_response_off": "\"state\": \"off\""
        }
      ]
    }
  }
}
```

Dependencies:
    requests==2.18.4
"""

import requests
from requests.auth import HTTPBasicAuth, HTTPDigestAuth
from typing import Union  # noqa

from fauxmo.plugins import FauxmoPlugin


class RESTAPIPlugin(FauxmoPlugin):
    """REST API plugin class.

    The Fauxmo class expects plugins to be instances of objects that have on()
    and off() methods that return True on success and False otherwise. This
    class takes a mix of url, method, header, body, and auth data and makes
    REST calls to a device.
    """

    def __init__(
        self,
        *,
        auth_type: str = None,
        headers: dict = None,
        method: str = "GET",
        name: str,
        off_cmd: str,
        off_data: dict = None,
        off_json: dict = None,
        on_cmd: str,
        on_data: dict = None,
        on_json: dict = None,
        password: str = None,
        port: int,
        state_cmd: str = None,
        state_data: dict = None,
        state_json: dict = None,
        state_method: str = "GET",
        state_response_off: str = None,
        state_response_on: str = None,
        user: str = None,
    ) -> None:
        """Initialize a RESTAPIPlugin instance.

        Args:
            auth_type: Either `basic` or `digest` if needed
            headers: Additional headers for both `on()` and `off()`
            method: HTTP method to be used for both `on()` and `off()`
            off_cmd: URL to be called when turning device off
            off_data: HTTP body to turn device off
            off_json: JSON body to turn device off
            on_cmd: URL to be called when turning device on
            on_data: HTTP body to turn device on
            on_json: JSON body to turn device on
            password: Password for `auth`
            state_cmd: URL to be called to determine device state
            state_data: Optional POST data to query device state
            state_json: Optional json data to query device state
            state_method: HTTP method to be used for `get_state()`
            state_response_off: If this string is in the response to state_cmd,
                                the device is off.
            state_response_on: If this string is in the response to state_cmd,
                               the device is on.
            user: Username for `auth`
        """
        self.method = method
        self.state_method = state_method
        self.headers = headers
        self.auth: Union[HTTPBasicAuth, HTTPDigestAuth] = None

        self.on_cmd = on_cmd
        self.off_cmd = off_cmd
        self.state_cmd = state_cmd

        self.on_data = on_data
        self.off_data = off_data
        self.state_data = state_data

        self.on_json = on_json
        self.off_json = off_json
        self.state_json = state_json

        self.state_response_on = state_response_on
        self.state_response_off = state_response_off

        if auth_type:
            if auth_type.lower() == "basic":
                self.auth = HTTPBasicAuth(user, password)

            elif auth_type.lower() == "digest":
                self.auth = HTTPDigestAuth(user, password)

        super().__init__(name=name, port=port)

    def on(self) -> bool:
        """Turn device on."""
        return self.set_state(self.on_cmd, self.on_data, self.on_json)

    def off(self) -> bool:
        """Turn device off."""
        return self.set_state(self.off_cmd, self.off_data, self.off_json)

    def set_state(self, cmd: str, data: dict, json: dict) -> bool:
        """Call HTTP method via Requests."""
        r = requests.request(
            self.method,
            cmd,
            data=data,
            json=json,
            headers=self.headers,
            auth=self.auth,
        )
        return r.status_code in [200, 201]

    def get_state(self) -> str:
        """Get device state.

        Returns:
            "on", "off", or "unknown"

        """
        if self.state_cmd is None:
            return "unknown"

        resp = requests.request(
            self.state_method,
            self.state_cmd,
            data=self.state_data,
            json=self.state_json,
            headers=self.headers,
            auth=self.auth,
        )

        if self.state_response_off in resp.text:
            return "off"
        elif self.state_response_on in resp.text:
            return "on"
        return "unknown"
